python/ray/data/_internal/planner/plan_write_op.py

```diff
@@ -51,8 +72,13 @@ def fn(blocks: Iterator[Block], ctx) -> Iterator[Block]:
         # type.
         import pandas as pd
 
-        write_result = WriteResult(num_rows=total_num_rows, size_bytes=total_size_bytes)
-        block = pd.DataFrame({"write_result": [write_result]})
+        block = pd.DataFrame(
+            {
+                "num_rows": [total_num_rows],
+                "size_bytes": [total_size_bytes],
+                "write_task_result": [ctx.kwargs.get("_data_sink_custom_result", None)],
```

Comment:
Rather than passing information through the `TaskContext`, can we directly yield the write returns and stats in `generate_write_fn.fn`?

```
# Pseudocode for `generate_write_fn.fn`
for block in blocks:
    write_return = datasink.write(block)
    yield Block({"write_return": write_return, "num_rows": block.num_rows()})
```

Reply:
Wait, on second thought, do we even still need `generate_collect_write_stats_fn`?

I think we can return the write return and statistics from `generate_write_fn.fn`, and then aggregate the statistics and create `WriteResult` on the driver?

Reply:
having 2 separate TransformFns allows optimization rules to insert certain operations in between them. And to pass data between them, TaskContext is probably the best place.